Feature/moon routerdealer step1

m3d opened the pull request:
This is the first step of #527 replacement - it provides the new callback functionality in `controller.py` but allows backward compatibility with `blocking=true` (which will be removed at the end). Also the communication between request/response modules is now using extra token.

m3d commented:
@zwn - another suspicious test failure:
```
======================================================================
FAIL: test_sigint_shell (osgar.test_record.RecorderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/osgar/osgar/osgar/test_record.py", line 125, in test_sigint_shell
    self.assertEqual(len(stderr), 2, stderr)
AssertionError: 1 != 2 : [b'2020-07-02 20:30:31,848 __main__         INFO     /home/runner/work/osgar/osgar/tmp78hcyyxc-200702_203031.log']

----------------------------------------------------------------------
Ran 244 tests in 19.933s

FAILED (failures=1)
##[error]Process completed with exit code 1.
```

zwn commented:
@m3d The test checks whether "INFO     committing suicide by SIGINT" is output to stderr. In this case, it was not. As to why - I have no idea :worried:.

zwn commented:
I am not sure what the goal of the PR is. First I thought the point is to allow `callback` to be called when the response is received but the `on_response` is never called anywhere (not even in tests). There is still the `print(dt, 'ignoring', channel)` worrying line. So - what is the one-sentence-pitch for this PR?

zwn commented:
As for the failed test - we might be sending the `SIGINT` too soon - before the handler is installed (there is the dreaded `time.sleep` synchronization anti-pattern). There is no immediate way to know when the signal handler is installed - ideas welcomed.

m3d commented:
> I am not sure what the goal of the PR is. First I thought the point is to allow `callback` to be called when the response is received but the `on_response` is never called anywhere (not even in tests). There is still the `print(dt, 'ignoring', channel)` worrying line. So - what is the one-sentence-pitch for this PR?

The goal of this PR is not to break existing functionality (which is using blocking read and prints "ignoring" messages) and allow usage of callbacks (must be called with blocking=False ... I was not sure if to change while cycle to `while callback is None and blocking`, so once you use callback (new way) then blocking is ignored).

m3d commented:
good point with the tests ... it was surely broken as the `on_<something>` has different parameters now. I also changed functionality to ignore blocking if `callback` is defined.

I was a bit surprised that inner function does not see/write to function variable `echo_data` ... so I changed it to global.

m3d commented:
I am going to merge this and potential corrections apply in following PR "step2".

zwn commented:
> I was a bit surprised that inner function does not see/write to function variable `echo_data` ... so I changed it to global.

It needs to be marked `nonlocal`.
